app/stats/patterns/creational/graph_abstract_factory_pattern.py

```diff
@@ -0,0 +1,161 @@
+"""
+*Abstract Factory Pattern for the main graph
+This pattern serves to provide an interface for different implementations of 
+time scopes, such as annually, monthly and daily.
+
+The implementations return a dictionary for the Altair library to use.
+"""
+
+from typing import Type
+
+import altair as alt
+import duckdb
+
+
+class Graph:
+    def __init__(self, start_year, end_year, tld) -> None:
+        self.start_year = start_year
+        self.end_year = end_year
+        self.tld = tld
+
+    def create_visualization(self) -> None:
+        raise NotImplementedError
+    
+    def get_source(self, query) -> None:
+        conn = duckdb.connect()
+        source = conn.execute(query).df()
+
+        # Convert the numbers of lookups to a more readable size
+        source['look_ups'] = source['look_ups']/1000000
+        return source
+
+class Annual(Graph):
+    def __init__(self, start_year, end_year, tld) -> None:
+        super().__init__(start_year, end_year, tld)
+
+        dataset_location = './stats/static/datasets/annual.development.parquet'
+
+        if (self.tld and self.tld != 'global'):
+            # Query based on Top-Level Domain (TLD)
+            query = "SELECT * FROM '{}' WHERE year >= {} AND year <= {} AND tld = '{}'".format(dataset_location, self.start_year, self.end_year, self.tld)
+            graph_title = "Greencheck annual overview for .{} - {} to {}".format(self.tld, self.start_year, self.end_year)
+        else:
+            # No TLD specified, so query all TLDs
+            query = "SELECT year, green, sum(look_ups) AS look_ups FROM '{}' WHERE year >= {} AND year <= {} GROUP BY year, green".format(dataset_location, self.start_year, self.end_year)

```

Comment:
Can you please use the DB API here?

I see why you might want to do it this way, but I think it would introduce a vulnerability.

If someone passed in something like `; DROP TABLE X` as their tld, we'd feel a bit sheepish, and I didn't see any data checking earlier in the request flow.

duckdb implements the same DB API as sqlite3 does with `execute` - which from memory exists to protect against these attacks.
https://docs.python.org/3.10/library/sqlite3.html#sqlite3.Cursor.execute

I'm not 100% on that - please check to see if does, and let me know otherwise. If it doesn't, we'd likely need different safeguard, as user input as a rule can not be trusted.

example of the DB API in duckdb:
https://github.com/duckdb/duckdb/blob/master/examples/python/duckdb-python.py#L18